apps/emqx/src/emqx_channel.erl

```diff
@@ -2088,6 +2111,45 @@ is_clientid_namespaced(ClientId, Tns) when is_binary(Tns) andalso Tns =/= <<>> -
 is_clientid_namespaced(_ClientId, _Tns) ->
     false.
 
+%%--------------------------------------------------------------------
+%% Adjust limiter
+
+-define(PT_LIMITER_ADJUST_KEY, {?MODULE, limiter_adjust_fn}).
+-spec set_limiter_adjustment_fn(fun((map()) -> ignore | {ok, emqx_limiter_client_container:t()})) ->
+    ok.
+set_limiter_adjustment_fn(Fn) when is_function(Fn, 1) ->
+    persistent_term:put(?PT_LIMITER_ADJUST_KEY, Fn).
+
+-spec get_limiter_adjustment_fn() ->
+    undefined | fun((map()) -> ignore | {ok, emqx_limiter_client_container:t()}).
+get_limiter_adjustment_fn() ->
+    persistent_term:get(?PT_LIMITER_ADJUST_KEY, undefined).
+
+-spec unset_limiter_adjustment_fn() -> ok.
+unset_limiter_adjustment_fn() ->
+    _ = persistent_term:erase(?PT_LIMITER_ADJUST_KEY),
+    ok.
+-undef(PT_LIMITER_ADJUST_KEY).
+
+adjust_limiter(_ConnPkt, #channel{clientinfo = ClientInfo} = Channel0) ->
+    maybe
+        Fn = get_limiter_adjustment_fn(),
+        true ?= is_function(Fn, 1),
+        #{zone := Zone, listener := ListenerId} = ClientInfo,
+        Tns = get_tenant_namespace(ClientInfo),
+        Context = #{zone => Zone, listener_id => ListenerId, tns => Tns},
```

Comment:
Nit

The `adjust` hook naming is quite generic, but the context content somewhat discovers that it is a tenant-specific limiter.
Also, we have two places for limiter creation.

Maybe
* call here a (new) regular fold hook (`emqx_hooks` based)
* pass the default (zone-listener) limiter as the default acc to it?

Thus we also won't need an additional hooking mechanism.

Reply:
> The adjust hook naming is quite generic, but the context content somewhat discovers that it is a tenant-specific limiter.

The _current context_ only needs this, so it doesn't imply it's _only_ for tenant stuff.  More stuff may be added to the context in the future, as the need arises.  It's useful to use a map for this, since they are extensible.

The problem I see with creating the limiter only at this point is that we'll have to create an "inconsistent" `#channel{}` record with `quota = undefined`, until it reaches this point of the processing pipeline.  I wanted to have _something_ there until we have enough information to potentially change it.  :thinking: